airflow/models/dagrun.py

```diff
@@ -596,19 +596,18 @@ def recalculate(self) -> _UnfinishedStates:
                         unfinished = unfinished.recalculate()
 
         leaf_task_ids = {t.task_id for t in dag.leaves}
-        leaf_tis = [ti for ti in tis if ti.task_id in leaf_task_ids if ti.state != TaskInstanceState.REMOVED]
-        # TODO: Remove 'getattr' if setup/teardown is available for mapped task
+        leaf_tis = {ti for ti in tis if ti.task_id in leaf_task_ids if ti.state != TaskInstanceState.REMOVED}
         if dag.teardowns:
             # when on_failure_fail_dagrun is `False`, the final state of the DagRun
             # will be computed as if the teardown task simply didn't exist.
-            teardown_task_ids = [t.task_id for t in dag.teardowns]
-            upstream_of_teardowns = [t.task_id for t in dag.tasks_upstream_of_teardowns]
-            teardown_tis = [ti for ti in tis if ti.task_id in teardown_task_ids]
-            on_failure_fail_tis = [ti for ti in teardown_tis if getattr(ti.task, "on_failure_fail_dagrun")]
-            tis_upstream_of_teardowns = [ti for ti in tis if ti.task_id in upstream_of_teardowns]
-            leaf_tis = list(set(leaf_tis) - set(teardown_tis))
-            leaf_tis.extend(on_failure_fail_tis)
-            leaf_tis.extend(tis_upstream_of_teardowns)
+            teardown_task_ids = {t.task_id for t in dag.teardowns}
+            upstream_of_teardowns = {t.task_id for t in dag.tasks_upstream_of_teardowns}
+            teardown_tis = {ti for ti in tis if ti.task_id in teardown_task_ids}
+            on_failure_fail_tis = {ti for ti in teardown_tis if ti.task.on_failure_fail_dagrun}
+            tis_upstream_of_teardowns = {ti for ti in tis if ti.task_id in upstream_of_teardowns}
+            leaf_tis -= teardown_tis
+            leaf_tis |= on_failure_fail_tis
+            leaf_tis |= tis_upstream_of_teardowns
```

Comment:
```suggestion
            leaf_tis.difference_update(ti for ti in tis if ti.task_id in teardown_task_ids)
            leaf_tis.update(
                (ti for ti in teardown_tis if ti.task.on_failure_fail_dagrun),
                (ti for ti in tis if ti.task_id in upstream_of_teardowns),
            )
```

Should be even faster by avoid creating extra sets.

Reply:
thanks @uranusjr . i assumed that the operators did the same thing under the hood.

i looked at the docs [here](https://docs.python.org/3.8/library/stdtypes.html#set-types-set-frozenset) and that's at least what it looks like... they have -= and difference_update defined together, same with |= and update. is there some underlying implementation difference? (can't see the source code unfortunately since it looks like it's in C)

Reply:
The operators _do_ the same under the hood, the difference here is I use generators instead of sets. But the operators don’t support those, only functions do.

> Note, the non-operator versions of the [update()](https://docs.python.org/3.8/library/stdtypes.html#frozenset.update), [intersection_update()](https://docs.python.org/3.8/library/stdtypes.html#frozenset.intersection_update), [difference_update()](https://docs.python.org/3.8/library/stdtypes.html#frozenset.difference_update), and [symmetric_difference_update()](https://docs.python.org/3.8/library/stdtypes.html#frozenset.symmetric_difference_update) methods will accept any iterable as an argument.

Reply:
ah yes makes sense.... maybe i will keep the existing variables (for readability) but convert them to generators and use the non-operators

Reply:
mrrr cancel that

Reply:
ok yeah, turns out a combination was required

`teardown_tis` was used two times, so it needed to be a set
the other two can be generator (and i did assign both of them to a variable which i think is helpful from readability perspective and should have same performance)